pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test: TestLogic_mixed_version_database_role_settings_role_id failed

pkg/sql/logictest/tests/cockroach-go-testserver-upgrade-to-master/cockroach-go-testserver-upgrade-to-master_test.TestLogic_mixed_version_database_role_settings_role_id [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17817516?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/17817516?buildTab=artifacts#/) on release-23.2 @ [b2b6beae048d7f8f88c77bd76fe9a75fd0fd29cc](https://github.com/cockroachdb/cockroach/commits/b2b6beae048d7f8f88c77bd76fe9a75fd0fd29cc):


```
        	GOROOT/src/os/exec/exec.go:717 +0x34
        created by os/exec.(*Cmd).Start in goroutine 14
        	GOROOT/src/os/exec/exec.go:716 +0x7bc
        Leaked goroutine: goroutine 30 [IO wait, 1 minutes]:
        internal/poll.runtime_pollWait(0xffff89810488, 0x72)
        	GOROOT/src/runtime/netpoll.go:343 +0xa0
        internal/poll.(*pollDesc).wait(0x4001bec120?, 0x40019b6000?, 0x1)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x28
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0x4001bec120, {0x40019b6000, 0x8000, 0x8000})
        	GOROOT/src/internal/poll/fd_unix.go:164 +0x200
        os.(*File).read(...)
        	GOROOT/src/os/file_posix.go:29
        os.(*File).Read(0x40039a0018, {0x40019b6000?, 0x4001134138?, 0x0?})
        	GOROOT/src/os/file.go:118 +0x70
        io.copyBuffer({0xffff3dfa3738, 0x4001134138}, {0x63c19a0, 0x40039a0018}, {0x0, 0x0, 0x0})
        	GOROOT/src/io/io.go:430 +0x168
        io.Copy(...)
        	GOROOT/src/io/io.go:389
        os/exec.(*Cmd).writerDescriptor.func1()
        	GOROOT/src/os/exec/exec.go:560 +0x44
        os/exec.(*Cmd).Start.func2(0x0?)
        	GOROOT/src/os/exec/exec.go:717 +0x34
        created by os/exec.(*Cmd).Start in goroutine 14
        	GOROOT/src/os/exec/exec.go:716 +0x7bc
        Leaked goroutine: goroutine 31 [IO wait]:
        internal/poll.runtime_pollWait(0xffff89810390, 0x72)
        	GOROOT/src/runtime/netpoll.go:343 +0xa0
        internal/poll.(*pollDesc).wait(0x4001bec1e0?, 0x40019be000?, 0x1)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:84 +0x28
        internal/poll.(*pollDesc).waitRead(...)
        	GOROOT/src/internal/poll/fd_poll_runtime.go:89
        internal/poll.(*FD).Read(0x4001bec1e0, {0x40019be000, 0x8000, 0x8000})
        	GOROOT/src/internal/poll/fd_unix.go:164 +0x200
        os.(*File).read(...)
        	GOROOT/src/os/file_posix.go:29
        os.(*File).Read(0x40039a0030, {0x40019be000?, 0x4001134150?, 0x0?})
        	GOROOT/src/os/file.go:118 +0x70
        io.copyBuffer({0xffff3dfa3738, 0x4001134150}, {0x63c19a0, 0x40039a0030}, {0x0, 0x0, 0x0})
        	GOROOT/src/io/io.go:430 +0x168
        io.Copy(...)
        	GOROOT/src/io/io.go:389
        os/exec.(*Cmd).writerDescriptor.func1()
        	GOROOT/src/os/exec/exec.go:560 +0x44
        os/exec.(*Cmd).Start.func2(0x0?)
        	GOROOT/src/os/exec/exec.go:717 +0x34
        created by os/exec.(*Cmd).Start in goroutine 14
        	GOROOT/src/os/exec/exec.go:716 +0x7bc
--- FAIL: TestLogic_mixed_version_database_role_settings_role_id (69.23s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_mixed_version_database_role_settings_role_id.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-44647

--- a/pkg/util/leaktest/leaktest.go
+++ b/pkg/util/leaktest/leaktest.go
@@ -136,11 +136,17 @@ func AfterTest(t T) func() {
 		// to see if the leak detector should be disabled for future tests.
 		if f, ok := t.(interface {
 			Failed() bool
-		}); ok && f.Failed() {
-			if err := diffGoroutines(orig); err != nil {
-				atomic.StoreUint32(&leakDetectorDisabled, 1)
+			Skipped() bool
+		}); ok {
+			switch {
+			case f.Failed():
+				if err := diffGoroutines(orig); err != nil {
+					atomic.StoreUint32(&leakDetectorDisabled, 1)
+				}
+				fallthrough
+			case f.Skipped():
+				return
 			}
-			return
 		}
 
 		if tb, ok := t.(testing.TB); ok {

